sdk/core/azure-core/azure/core/pipeline/policies/_authentication.py

```diff
@@ -172,6 +252,41 @@ def on_exception(self, request):
         return
 
 
+class MultitenantTokenCredentialPolicy(BearerTokenCredentialPolicy):
+    """Adds a bearer token Authorization header to requests, for the tenant provided in authorization challenges.
+
+    See https://docs.microsoft.com/azure/active-directory/develop/claims-challenge for documentation on AAD
+    authentication challenges.
+
+    :param credential: The credential.
+    :type credential: ~azure.core.TokenCredential
+    :param str scopes: Lets you specify the type of access needed.
+    :raises: :class:`~azure.core.exceptions.ServiceRequestError`
+    """
+
+    def on_challenge(self, request: "PipelineRequest", response: "PipelineResponse") -> bool:
+        """Authorize request according to an authentication challenge
+
+        This method is called when the resource provider responds 401 with a WWW-Authenticate header.
+
+        :param ~azure.core.pipeline.PipelineRequest request: the request which elicited an authentication challenge
+        :param ~azure.core.pipeline.PipelineResponse response: the resource provider's response
+        :returns: a bool indicating whether the policy should send the request
+        """
+        try:
+            challenge = _HttpChallenge(request.http_request.url, response.http_response.headers.get("WWW-Authenticate"))
+            # azure-identity credentials require an AADv2 scope but the challenge may specify an AADv1 resource
+            scope = challenge.get_scope() or challenge.get_resource() + "/.default"
+        except ValueError:
+            return False
+
+        body = request.context.pop("key_vault_request_data", None)
```

Comment:
Is it only for key vault?

Reply:
This is a general-purpose policy that will work for Key Vault, Tables, and any other service that opts in to challenge authentication with tenant discovery. This KV-specific bit is coming from the current KV implementation, but I may be able to remove this if it turns out to be unnecessary. For non-KV scenarios it wouldn't cause an error, and would just be superfluous.

Reply:
Update: this did turn out to be unnecessary, and was part of KV's strategy to remove request information before intentionally triggering a 401. The original body content would be removed and stored in the request context, and then brought back in for the authenticated request. Since we're no longer making an initial, empty request to trigger 401s, we can remove this.